File: S7/Desarrollo_casos/Dijsktra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <cstdlib>
#include <ctime>
using namespace std;

class Grafo
{
    int n;
    vector<vector<int>> adyacentes;
    vector<vector<int>> pesos;

public:
    Grafo(int n){
        this->n = n;
        this->adyacentes.resize(n);
        this->pesos.resize(n);
        for (int i = 0; i < n; ++i) {
            this->adyacentes[i].resize(n);
            this->pesos[i].resize(n);
        }
    }

    void agregar_arista(int u, int v, int peso)
    {
        this->adyacentes[u].push_back(v);
        this->pesos[u][v] = peso;
    }

    vector<int> obtener_adyacentes(int nodo)
    {
        return this->adyacentes[nodo];
    }

    int obtener_peso(int u, int v)
    {
        if(this->pesos[u][v]){
            return this->pesos[u][v];
        }
        return numeric_limits<int>::max();
    }

    int obtener_n()
    {
        return this->n;
    }
};

class Tupla
{
public:
    vector<int> distancias;
    vector<int> previo;
    vector<bool> visitado;

    Tupla(int n)
    {
        this->distancias.resize(n, numeric_limits<int>::max());
        this->previo.resize(n, -1);
        this->visitado.resize(n, false);
    }
};

class ColaPrioridad
{
    typedef pair<int, int> pii;
    priority_queue<pii, vector<pii>, greater<pii>> pq;

public:
    void insertar(int elemento, int prioridad)
    {
        pq.push({prioridad, elemento});
    }

    int extraer_minimo()
    {
        int elemento = pq.top().second;
        pq.pop();
        return elemento;
    }

    void insertar_o_decrementar(int elemento, int nueva_prioridad)
    {
        pq.push({nueva_prioridad, elemento});
    }

    bool vacia()
    {
        return pq.empty();
    }
};

Tupla *Dijkstra(Grafo *G, int origen)
{
    Tupla *solucion = new Tupla(G->obtener_n());

    solucion->distancias[origen] = 0;
    ColaPrioridad *cola_prioridad = new ColaPrioridad();
    cola_prioridad->insertar(origen, 0);

    while (!cola_prioridad->vacia())
    {
        int actual = cola_prioridad->extraer_minimo();
        solucion->visitado[actual] = true;

        for (int vecino : G->obtener_adyacentes(actual))
        {
            if (!solucion->visitado[vecino])
            {
                int costo = G->obtener_peso(actual, vecino);
                int nueva_distancia = solucion->distancias[actual] + costo;

                if (nueva_distancia < solucion->distancias[vecino])
                {
                    solucion->distancias[vecino] = nueva_distancia;
                    solucion->previo[vecino] = actual;
                    cola_prioridad->insertar_o_decrementar(vecino, nueva_distancia);
                }
            }
        }
    }

    delete cola_prioridad;
    return solucion;
}

void insertar_dinamico(int n){
    clock_t start_time = clock(); 
    srand(time(nullptr));
    
    Grafo G = Grafo(n);
    int v = n - 1;
    int nodo_destino = rand( )% (v + 1);
    int peso = rand() % 10 + 1;
    G.agregar_arista(0, nodo_destino, peso);

    for (int i = 0; i < n-1; i++){  
        int nodo_origen = rand() % (v + 1);
        peso = rand() % 10 + 1;
        nodo_destino = rand( )% (v + 1);
        while (nodo_origen == nodo_destino)
        {
            nodo_destino = rand( )% (v + 1);
        }
        G.agregar_arista(nodo_origen, nodo_destino, peso);
    }

    clock_t end_time = clock();
    double elapsed_time = double(end_time - start_time) / CLOCKS_PER_SEC;

    int ORIGEN = 0;
    Tupla *solucion = Dijkstra(&G, ORIGEN);

    cout << "Distancias mínimas desde el nodo " << ORIGEN << ":" << endl;
    for (int i = 0; i < solucion->distancias.size(); ++i)
    {
        cout << "Nodo " << i << ": " << solucion->distancias[i] << endl;
    }

    cout << "Tiempo promedio para n = " << n << ": " << elapsed_time << " segundos" << endl;

    delete solucion;
}

int main()
{
    srand(time(nullptr));
    insertar_dinamico(10);
    return 0;
}

File: S7/Desarrollo_casos/Dijsktra.py
import heapq


class Grafo:
    def __init__(self, n):
        self.n = n
        self.adyacentes = [[] for _ in range(n)]
        self.pesos = {}

    def agregar_arista(self, u, v, peso):
        self.adyacentes[u].append(v)
        self.pesos[(u, v)] = peso

    def obtener_adyacentes(self, nodo):
        return self.adyacentes[nodo]

    def obtener_peso(self, u, v):
        return self.pesos.get((u, v), float('inf'))


class ColaPrioridad:
    def __init__(self):
        self.elementos = []

    def insertar(self, elemento, prioridad):
        heapq.heappush(self.elementos, (prioridad, elemento))

    def extraer_minimo(self):
        return heapq.heappop(self.elementos)[1]

    def insertar_o_decrementar(self, elemento, nueva_prioridad):
        self.eliminar(elemento)
        self.insertar(elemento, nueva_prioridad)

    def eliminar(self, elemento):
        indice = self.encontrar_elemento(elemento)
        if indice != -1:
            self.elementos.pop(indice)
            heapq.heapify(self.elementos)

    def encontrar_elemento(self, elemento):
        for i, (_, elem) in enumerate(self.elementos):
            if elem == elemento:
                return i
        return -1


def dijkstra_code(G: Grafo, origen):
    distancias = [float('inf')] * G.n
    previo = [None] * G.n
    visitado = [False] * G.n

    distancias[origen] = 0
    cola_prioridad = ColaPrioridad()
    cola_prioridad.insertar(origen, 0)

    while cola_prioridad.elementos:
        actual = cola_prioridad.extraer_minimo()
        visitado[actual] = True

        for vecino in G.obtener_adyacentes(actual):
            if visitado[vecino] is False:
                costo = G.obtener_peso(actual, vecino)
                nueva_distancia = distancias[actual] + costo

                if nueva_distancia < distancias[vecino]:
                    distancias[vecino] = nueva_distancia
                    previo[vecino] = actual
                    cola_prioridad.insertar_o_decrementar(
                        vecino, nueva_distancia)
    return distancias, previo


# Crear un grafo de prueba
grafo = Grafo(5)
grafo.agregar_arista(0, 1, 10)
grafo.agregar_arista(0, 2, 5)
grafo.agregar_arista(1, 2, 2)
grafo.agregar_arista(1, 3, 1)
grafo.agregar_arista(2, 1, 3)
grafo.agregar_arista(2, 3, 9)
grafo.agregar_arista(2, 4, 2)
grafo.agregar_arista(3, 4, 4)
grafo.agregar_arista(4, 3, 6)

# Ejecutar el algoritmo de Dijkstra desde el nodo 0
ORIGEN = 0
distancias_1, previo_1 = dijkstra_code(grafo, ORIGEN)

# Imprimir las distancias mínimas desde el nodo de origen a todos los demás nodos
print("Distancias mínimas desde el nodo", ORIGEN, ":")
for j, distancia in enumerate(distancias_1):
    print(f"Nodo {j}: {distancia}")
